### ISSUE TITLE
[java] UnusedAssignment: False Negative when Assigning Field in Nested Class

### ISSUE THREAD

**Issue by mohui1999** [pmd/pmd#4636]
**Affects PMD Version:** 7.0.0-rc3

**Rule:**[UnusedAssignment](https://docs.pmd-code.org/latest/pmd_rules_java_bestpractices.html#unusedassignment)

**Description:**
When the `method` is placed inside the `NestedClass`, PMD fails to detect that the variable `bar` is being assigned a value and subsequently overwritten without being used. However, when the `method` is not inside the `NestedClass`, PMD can correctly detect and report the unused assignment for the variable `bar`. We consider the case with `NestedClass` to be a false negative.

**Code Sample demonstrating the issue:**
Maybe a false negative:

```java
public class SomeClass {
    private Boolean bar;

    class NestedClass {
        public void method(String s) {         
            bar = new Boolean(s);   // Not report UnusedAssignment
            bar = new Boolean("some arbitrary string is just false");   // Not report UnusedAssignment
            bar = Boolean.valueOf(s);
        }
    }
}
```

Correctly reported code:

```java

public class SomeClass {
      private Boolean bar;

      public void method(String s) {
            this.bar = new Boolean(s);       // report UnusedAssignment
            this.bar = new Boolean("some arbitrary string is just false");   // report UnusedAssignment
            this.bar = Boolean.valueOf(s);  
      }
}
```

**Expected outcome:**
```
UnusedAssignment:       The value assigned to field 'bar' is never used
```
PMD should report a violation at line 7,8, but doesn't. This is a false negative.

**Running PMD through:** *[CLI]*
